lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,406 @@
+            this.buildInFunctionProperties);
+          if (meaningfulProps.length > 0) {
+            // If this function object contains meaningful properties.
+            if (!map[interfaceName]) map[interfaceName] = [];
+            this.arrayMerge_(map[interfaceName],
+              this.getObjectProperties_(og, id),
+              this.getObjectProperties_(og, prototypeId));
+          }
+          if (og.getPrototype(id)) {
+            this.extractAPI_(og, map, og.getPrototype(id),
+              `${name}.__proto__`, visited, {proto: true});
+          }
+        } else if (!og.isFunction(id)) {
+          // Object is not a first level object or prototype function,
+          // nor is it a function.
+          let ownProperty = this.getObjectProperties_(og, id);
+          if (this.setMinus_(ownProperty,
+            this.buildInObjectProperties).length > 0) {
+            let className = this.getClassName_(id, og);
+            if ((!className || className === 'Object') && firstLevel) {
+              // If object is at first level and its class name is "Object",
+              // this object is considered as an object library,
+              // such as MATH, console.
+              let prototypeId = og.lookup('prototype', id);
+              // In Firefox, Edge, (maybe some other vendors or versions),
+              // objects can have prototype properties.
+              let prototypeProps = [];
+              if (prototypeId) {
+                prototypeProps = this.getObjectProperties_(og, prototypeId);
+              }
+              if (!map[name]) map[name] = [];
+              this.arrayMerge_(map[name],
+                this.getObjectProperties_(og, id), prototypeProps);
+            } else {
+              // This is not a object library, try to add its property to
+              // it's prototype. For example, document.body.style contains
+              // meaningful properties of CSSStyleDeclaration, then its
+              // properties will be added to CSSStyleDeclaration interface.
+              if (!className || className === 'Object') return;
+              // Skip if its class is Object, or cannot find it's class name.
+              // We don't want to add extra properties for Object.
+              if (!map[className]) map[className] = [];
+              this.arrayMerge_(map[className],
+                this.getObjectProperties_(og, id));
+            }
+          }
+          if (og.getPrototype(id)) {
+            // Visit object __proto__ if its __proto__ exists.
+            this.extractAPI_(og, map, og.getPrototype(id),
+              `${name}.__proto__`, visited, {proto: true});
+          }
+        }
+
+        // Don't visit __proto__'s properties
+        // since __proto__ does not contain instances.
+        if (proto) return;
+
+        // Visit object's properties.
+        let propertiesIdMap = og.getPropertiesIds(id);
+        let keys = Object.keys(propertiesIdMap);
```

Comment:
Possible improvement: `Object.keys()-then-iterate` generally performs slower than `for (var key in o) { if (o.hasOwnProperty(key)) { ... } }`. Maybe try it out to see whether there's an appreciable difference. If not, don't bother changing this.